services/blockstorage/adapter/filesystem/file_system_persistence.go

```diff
@@ -247,26 +247,36 @@ func buildIndex(r io.Reader, firstBlockOffset int64, logger log.Logger, c blockC
 	return bhIndex, nil
 }
 
+//func validateCandidateBlockHeight(candidateBlockHeight primitives.BlockHeight, syncState internodesync.SyncState, logger log.Logger) error {
+//	if (syncState.LastSyncedHeight > syncState.InOrderHeight && candidateBlockHeight != syncState.LastSyncedHeight-1) ||
+//		(syncState.InOrderHeight == syncState.TopHeight && candidateBlockHeight <= syncState.InOrderHeight) {
+//		err := fmt.Errorf("trying to write a block with height (%d) which does not match current storage state (%v)", uint64(candidateBlockHeight), syncState)
+//		logger.Info(err.Error())
+//		return err
+//	}
+//	return nil
+//}
+
 func (f *BlockPersistence) WriteNextBlock(blockPair *protocol.BlockPairContainer) (bool, primitives.BlockHeight, error) {
 	f.blockWriter.Lock()
 	defer f.blockWriter.Unlock()
 
 	bh := getBlockHeight(blockPair)
-	lastWrittenHeight := f.bhIndex.lastSyncedHeight
-	// TODO: calc next height and avoid mapping
-	if _, found := f.bhIndex.fetchBlockOffset(bh); found { // block already in storage, ignores as defined in test expected
-		return false, lastWrittenHeight, nil
+
+	syncState := f.bhIndex.getSyncState()
+	if err := f.bhIndex.validateCandidateBlockHeight(bh); err != nil {
+		return false, syncState.LastSyncedHeight, nil
```

Comment:
shouldn't this be `inOrder`? ongoing sync sessions are supposed to be encapsulated and hidden from callers of the service (`CommitBlock` for instance).